// fp-ts/lib/Ring.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var Semiring_1 = require("./Semiring");
/** @function */
exports.getFunctionRing = function (ring) {
    return __assign({}, Semiring_1.getFunctionSemiring(ring), { sub: function (f) { return function (g) { return function (x) { return ring.sub(f(x))(g(x)); }; }; } });
};
/**
 * `negate x` can be used as a shorthand for `zero - x`
 * @function
 */
exports.negate = function (ring) { return function (a) {
    return ring.sub(ring.zero())(a);
}; };
/** @function */
exports.getProductRing = function (RA, RB) {
    return {
        add: function (_a) {
            var a1 = _a[0], b1 = _a[1];
            return function (_a) {
                var a2 = _a[0], b2 = _a[1];
                return [RA.add(a1)(a2), RB.add(b1)(b2)];
            };
        },
        zero: function () { return [RA.zero(), RB.zero()]; },
        mul: function (_a) {
            var a1 = _a[0], b1 = _a[1];
            return function (_a) {
                var a2 = _a[0], b2 = _a[1];
                return [RA.mul(a1)(a2), RB.mul(b1)(b2)];
            };
        },
        one: function () { return [RA.one(), RB.one()]; },
        sub: function (_a) {
            var a1 = _a[0], b1 = _a[1];
            return function (_a) {
                var a2 = _a[0], b2 = _a[1];
                return [RA.sub(a1)(a2), RB.sub(b1)(b2)];
            };
        }
    };
};
